### PR TITLE
Cnp 1283 cnp test

### PR DESCRIPTION
**Pull request by danielev04**
Notes:
- added MI support

### CODE REVIEW

**Comment by danielev04** [comment]
> no, upstream first and use config as code, this is not on.
azure-vm-agent-plugin configuration moved under cac (sandbox env)

https://github.com/hmcts/cnp-jenkins-config/pull/249/commits/93e6def8ae1fba99b2f8d62ad19063cb6682b11b

**Comment by danielev04** [comment]
> no, upstream first and use config as code, this is not on.

azure-vm-agent-plugin configuration moved under cac (sandbox env)
https://github.com/hmcts/cnp-jenkins-config/pull/249/commits/93e6def8ae1fba99b2f8d62ad19063cb6682b11b

File: cac-sandbox.yml [hunk 1]
```diff
@@ -29,7 +29,965 @@ jenkins:
...
+        doNotUseMachineIfInitFails: true
+        enableMSI: false
+        enableUAMI: false
+        executeInitScriptAsRoot: true
+        existingStorageAccountName: "mgmtvmimgstorecnptest"
+        imageReference:
+          id: "/subscriptions/bf308a5c-0624-4334-8ff8-8dca9fd43783/resourceGroups/cnp-vmimages-sandbox/providers/Microsoft.Compute/images/moj-centos-agent74-20181113164555"
+          version: "latest"
+        imageTopLevelType: "advanced"
+        initScript: "usermod -a -G docker jenkinsssh\numount /mnt/resource\nmkdir\
```
Comment (timja) [inline cac-sandbox.yml]: can you format the initScript for reading please:
should look like this:
https://github.com/hmcts/cnp-jenkins-config/blob/master/cac-test-local.yml#L26-L48

the library the configuration-as-code plugin uses doesn't support exporting multiline strings properly

File: playbook.yml [hunk 2]
```diff
@@ -146,6 +145,9 @@
       - file_name: 'sonar.hpi'
         checksum: 'sha256:03e2372d36c5bcd7794e0ccafc89bdb552f68ab905f82a55ebfd2e1f201d602f'
         url: 'https://github.com/hmcts/sonarqube-plugin/releases/download/sonar-2.9-hmcts/sonar.hpi'
+      - file_name: 'azure-vm-agents.hpi'
+        checksum: 'sha256:99e4c0754a4c5357d9788d11d4335d40fe115a5c9d3dde3ecc7e074ba38f7059'
+        url: 'https://github.com/danielev04/azure-vm-agents-plugin/releases/download/1.0.2-MI-20190606.3/azure-vm-agents.hpi'
```
Comment (timja) [inline playbook.yml]: this can be retrieved from the jenkins incrementals maven repo:
https://repo.jenkins-ci.org/incrementals/org/jenkins-ci/plugins/azure-vm-agents/1.0.2-rc686.884e8965006c/azure-vm-agents-1.0.2-rc686.884e8965006c.hpi

File: cac-sandbox.yml (hunk 1 above)
Comment (danielev04) [reply cac-sandbox.yml]: done

File: playbook.yml (hunk 2 above)
Comment (danielev04) [reply playbook.yml]: done

File: roles/jenkins/tasks/azure-agents.yml [hunk 3]
```diff
@@ -1,10 +1,10 @@
 ---
 
-- name: Install Jenkins credential using script for the {{ credential.id }} secret.
+- name: Configure Azure Agent VM - generic profile used with legacy SPs login.
```
Comment (timja) [inline roles/jenkins/tasks/azure-agents.yml]: is this change needed?

File: cac-sandbox.yml [hunk 4]
```diff
@@ -29,7 +29,1160 @@ jenkins:
             value: DCD-CFTAPPS-SBOX
           - key: AKS_AAT_SUBSCRIPTION_NAME
             value: DCD-CFTAPPS-SBOX
-
+  clouds:
```
Comment (timja) [inline cac-sandbox.yml]: this file shouldn't be changed until the plugin has been upgraded on sandbox

File: cac-sandbox.yml [hunk 5]
```diff
@@ -29,7 +29,1160 @@ jenkins:
...
+  - azureVM:
+      azureCredentialsId: "jenkinsServicePrincipal"
+      cloudName: "cnp-azure"
+      deploymentTimeout: 1200
+      existingResourceGroupName: "mgmt-vmimg-store-cnptest"
```
Comment (timja) [inline cac-sandbox.yml]: shouldn't this be sandbox not cnp test?

File: cac-sandbox.yml [hunk 6]
```diff
@@ -29,7 +29,1160 @@ jenkins:
...
+      - agentLaunchMethod: "SSH"
+        agentWorkspace: "/opt/jenkins"
+        builtInImage: "Windows Server 2016"
+        credentialsId: "vm_agent_creds"
+        diskType: "managed"
+        doNotUseMachineIfInitFails: true
+        enableMSI: false
+        enableUAMI: false
+        executeInitScriptAsRoot: true
+        existingStorageAccountName: "mgmtvmimgstorecnptest"
```
Comment (timja) [inline cac-sandbox.yml]: sandbox?

File: cac-sandbox.yml [hunk 7]
```diff
@@ -29,7 +29,1160 @@ jenkins:
...
+        imageReference:
+          id: "/subscriptions/bf308a5c-0624-4334-8ff8-8dca9fd43783/resourceGroups/cnp-vmimages-sandbox/providers/Microsoft.Compute/images/moj-centos-agent74-20181113164555"
```
Comment (timja) [inline cac-sandbox.yml]: pretty sure this is the old agent version?
https://github.com/hmcts/cnp-infrastructure-bootstrap/blob/master/sandbox.tfvars#L5

File: cac-sandbox.yml [hunk 8]
```diff
@@ -29,7 +29,1160 @@ jenkins:
...
+        imageTopLevelType: "advanced"
+        initScript: | 
+          usermod -a -G docker jenkinsssh
+          umount /mnt/resource
+          mkdir -pv /opt/jenkins
+          mount /dev/sdc1 /opt/jenkins
+          chown -R jenkinsssh:jenkinsssh /opt/jenkins
+          mv /tmp/jenkinsssh_id_rsa /home/jenkinsssh/.ssh/id_rsa
+          chown jenkinsssh:jenkinsssh /home/jenkinsssh/.ssh/id_rsa
+          chmod 0600 /home/jenkinsssh/.ssh/id_rsa\n\
```
Comment (timja) [inline cac-sandbox.yml]:
```suggestion
          chmod 0600 /home/jenkinsssh/.ssh/id_rsa
```

File: cac-sandbox.yml [hunk 9]
```diff
@@ -29,7 +29,1160 @@ jenkins:
...
+        templateDisabled: false
+        templateName: "cnp-jenkins-builders"
+        usageMode: "Use this node as much as possible"
+        usePrivateIP: true
+        virtualMachineSize: "Standard_D4_v3"
+        virtualNetworkName: "mgmt-infra-cnptest"
+        virtualNetworkResourceGroupName: "mgmt-infra-cnptest"
+  - azureVM:
+      azureCredentialsId: "jenkinsServicePrincipal"
+      cloudName: "cnp-azure-team-a"
```
Comment (timja) [inline cac-sandbox.yml]:
```suggestion
      cloudName: "cnp-azure-cnp"
```

File: cac-sandbox.yml [hunk 10]
```diff
@@ -29,7 +29,1160 @@ jenkins:
...
+        templateName: "cnp-jenkins-builders-corecasedata"
+        uamiID: "/subscriptions/bf308a5c-0624-4334-8ff8-8dca9fd43783/resourcegroups/managed-identities-sandbox-rg/providers/Microsoft.ManagedIdentity/userAssignedIdentities/ccd-sandbox-mi"
+        usageMode: "Only build jobs with label expressions matching this node"
+        usePrivateIP: true
+        virtualMachineSize: "Standard_D4_v3"
+        virtualNetworkName: "mgmt-infra-cnptest"
+        virtualNetworkResourceGroupName: "mgmt-infra-cnptest"
+  - azureVM:
+      azureCredentialsId: "jenkinsServicePrincipal"
+      cloudName: "cnp-azure-courtstribunalsservicecentres"
```
Comment (timja) [inline cac-sandbox.yml]:
```suggestion
      cloudName: "cnp-azure-ctsc"
```

File: cac-sandbox.yml [hunk 11]
```diff
@@ -29,7 +29,1160 @@ jenkins:
...
+        templateName: "cnp-jenkins-builders-courtstribunalsservicecentres"
+        uamiID: "/subscriptions/bf308a5c-0624-4334-8ff8-8dca9fd43783/resourcegroups/managed-identities-sandbox-rg/providers/Microsoft.ManagedIdentity/userAssignedIdentities/ctsc-sandbox-mi"
+        usageMode: "Only build jobs with label expressions matching this node"
+        usePrivateIP: true
+        virtualMachineSize: "Standard_D4_v3"
+        virtualNetworkName: "mgmt-infra-cnptest"
+        virtualNetworkResourceGroupName: "mgmt-infra-cnptest"
+  - azureVM:
+      azureCredentialsId: "jenkinsServicePrincipal"
+      cloudName: "cnp-azure-evidencemanagement"
```
Comment (timja) [inline cac-sandbox.yml]: does this look ok in the UI on the main page? worried its too long can you send a screenshot from a running instance please of the main page of jenkins

File: playbook.yml [hunk 12]
```diff
@@ -4,7 +4,7 @@
   tasks:
     - name: set additional vars
       set_fact:
-        resolved_env: "{{ (env == 'prod') | ternary('prod', 'sandbox') }}"
+        resolved_env: "test"
```
[the remark this answers is not in the document]
Comment (danielev04) [reply playbook.yml]: done

File: playbook.yml [hunk 13]
```diff
@@ -146,6 +145,9 @@
       - file_name: 'sonar.hpi'
         checksum: 'sha256:03e2372d36c5bcd7794e0ccafc89bdb552f68ab905f82a55ebfd2e1f201d602f'
         url: 'https://github.com/hmcts/sonarqube-plugin/releases/download/sonar-2.9-hmcts/sonar.hpi'
+      - file_name: 'azure-vm-agents-1.0.2-rc686.884e8965006c.hpi'
+        checksum: 'sha256:ebaee5cd5ead742d604c8bf29aea391fde4c49ce8f2ddb1fd0ecb296351ff43e'
+        url: 'https://repo.jenkins-ci.org/incrementals/org/jenkins-ci/plugins/azure-vm-agents/1.0.2-rc686.884e8965006c/azure-vm-agents-1.0.2-rc686.884e8965006c.hpi'
```
[the remark this answers is not in the document]
Comment (danielev04) [reply playbook.yml]: ok

File: roles/jenkins/tasks/azure-agents.yml (hunk 3 above)
Comment (danielev04) [reply roles/jenkins/tasks/azure-agents.yml]: yes

File: cac-sandbox.yml (hunk 4 above)
Comment (danielev04) [reply cac-sandbox.yml]: this version of the plugin https://github.com/hmcts/cnp-jenkins-config/blob/CNP-1283-CNPTest/playbook.yml#L148-L150 support cac

File: cac-sandbox.yml (hunk 5 above)
Comment (danielev04) [reply cac-sandbox.yml]: fixed

File: cac-sandbox.yml (hunk 7 above)
Comment (danielev04) [reply cac-sandbox.yml]: do you have a new one?

File: cac-sandbox.yml (hunk 8 above)
Comment (danielev04) [reply cac-sandbox.yml]: fixed

File: cac-sandbox.yml (hunk 9 above)
Comment (danielev04) [reply cac-sandbox.yml]: there's a team-a example repo and pipeline, for this reason, we need it

File: cac-sandbox.yml (hunk 10 above)
Comment (danielev04) [reply cac-sandbox.yml]: done

File: cac-sandbox.yml (hunk 11 above)
Comment (danielev04) [reply cac-sandbox.yml]: Screenshot 2019-06-10 at 12 58 48

File: cac-sandbox.yml (hunk 5 above)
Comment (danielev04) [reply cac-sandbox.yml]: fixed

File: playbook.yml (hunk 13 above)
[the remark this answers is not in the document]
Comment (timja) [reply playbook.yml]: the latest version is:
https://repo.jenkins-ci.org/incrementals/org/jenkins-ci/plugins/azure-vm-agents/1.0.2-rc691.d86f19c14fa7/azure-vm-agents-1.0.2-rc691.d86f19c14fa7.hpi
but likely to get merged today or tomorrow

File: roles/jenkins/tasks/azure-agents.yml (hunk 3 above)
Comment (timja) [reply roles/jenkins/tasks/azure-agents.yml]: this file should be removed with changing over to azure vm agent

File: cac-sandbox.yml (hunk 4 above)
Comment (timja) [reply cac-sandbox.yml]: cac-sandbox will get applied after this is merged, the plugin has to be upgraded before this PR can be merged

File: cac-sandbox.yml (hunk 7 above)
Comment (timja) [reply cac-sandbox.yml]: I linked to the newer / current version

File: playbook.yml (hunk 13 above)
[the remark this answers is not in the document]
Comment (danielev04) [reply playbook.yml]: updated

File: cac-sandbox.yml (hunk 7 above)
Comment (danielev04) [reply cac-sandbox.yml]: fixed

File: playbook.yml (hunk 13 above)
[the remark this answers is not in the document]
Comment (danielev04) [reply playbook.yml]: updated with azure-vm-agents-1.0.2-rc691.d86f19c14fa7.hpi

File: cac-sandbox.yml (hunk 4 above)
Comment (danielev04) [reply cac-sandbox.yml]: why the plugin has to be upgraded before this PR?

Comment (timja) [reply cac-sandbox.yml]: to decouple the release, we may (and do) need to be able to change the config at any time. if the plugin isn't updated first then the config won't be able to be applied

File: playbook.yml [hunk 14]
```diff
@@ -146,6 +145,9 @@
       - file_name: 'sonar.hpi'
         checksum: 'sha256:03e2372d36c5bcd7794e0ccafc89bdb552f68ab905f82a55ebfd2e1f201d602f'
         url: 'https://github.com/hmcts/sonarqube-plugin/releases/download/sonar-2.9-hmcts/sonar.hpi'
+      - file_name: 'azure-vm-agents-1.0.2-rc686.884e8965006c.hpi'
```
Comment (timja) [inline playbook.yml]: add a comment that this is merged on master pending release and should be updated after 1.0.2 is release

File: cac-sandbox.yml (hunk 5 above)
Comment (danielev04) [reply cac-sandbox.yml]: left only team-a

File: cac-sandbox.yml (hunk 4 above)
Comment (danielev04) [reply cac-sandbox.yml]: commented out 'clouds' section for now

File: playbook.yml (hunk 14 above)
Comment (danielev04) [reply playbook.yml]: done
